### ISSUE TITLE
TS2345 error when using function generics for React Components

### ISSUE THREAD

**Issue by luan-dang-techlabs** [microsoft/TypeScript#56604]
### 🔎 Search Terms

React TS2345 instantiated with a different subtype of constraint

### ⏯ Playground Link

https://www.typescriptlang.org/play?pretty=false&ts=4.9.5&ssl=24&ssc=1&pln=2&pc=2#code/FASwtgDg9gTgLgAgEoFMCGBjOAaBBvVTOAUQBsUwUA7HBAVwGcViAzFlLAXwRZijAQByGOiyCA3MFA0UMFphQIACnwgMAPABUEKAB5xqAEwb4EAbQZwYALgSWYIKgHMAurfuOnnAHx5gCAIQMNFJSACNMAGtbAAo0GCcGW00ASgQAXm8EADcoEENJTikMKCpLZQyELR19IxM8c3t3K083OxbncQRObG8YvGDQiIxIzlsVKDUtbxTbQiwyCmpETPwpQPomVnYsGJi01b8NwMHwqP6eEBRSQyo0SltBNEFulMkNnvMXN-9AkTg6DAqAgYr9jlUABYAJm8EOuEHUAHpod4wQgfkVgCw6FQsCBSggIWgqIZyABhEJnEZxBJJUwsK43O4PdoOTqvWy5fJrAIlMpQcgAOlIUCcNMSGKkwHUSiqeAZ11u9xQzTZXiyp2GkXSeCJJPJlK13ERqKkQA

### 💻 Code

```ts
import React, {ReactElement, useEffect} from 'react';

interface Props<T extends { [str: string]: string}>{
    callback: (args: T) => void;
}

const P = <T extends { [str: string]: string; },>({callback}: Props<T>): ReactElement => {

    useEffect(() => {
       callback({ fieldname: 'a' });
    }, []);
    return (
        <h2>help</h2>
    );
}

function handleCallback(args: { fieldname: string; }): void {
  console.log(args);
}


<P <{fieldname: string}> callback={handleCallback} />

```

### 🙁 Actual behavior

TypeScript is reporting an error.

```
Argument of type '{ fieldname: string; }' is not assignable to parameter of type 'T'.
  '{ fieldname: string; }' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{ [str: string]: string; }'.
```

<img width="1888" alt="Screenshot 2023-11-29 at 10 43 46 AM" src="https://github.com/microsoft/TypeScript/assets/91907117/9bc0fa5d-8106-4065-a3be-6c0f21cc2f14">

[example](
https://www.typescriptlang.org/play?pretty=false&ts=4.9.5&ssl=24&ssc=1&pln=2&pc=2#code/FASwtgDg9gTgLgAgEoFMCGBjOAaBBvVTOAUQBsUwUA7HBAVwGcViAzFlLAXwRZijAQByGOiyCA3MFA0UMFphQIACnwgMAPABUEKAB5xqAEwb4EAbQZwYALgSWYIKgHMAurfuOnnAHx5gCAIQMNFJSACNMAGtbAAo0GCcGW00ASgQAXm8EADcoEENJTikMKCpLZQyELR19IxM8c3t3K083OxbncQRObG8YvGDQiIxIzlsVKDUtbxTbQiwyCmpETPwpQPomVnYsGJi01b8NwMHwqP6eEBRSQyo0SltBNEFulMkNnvMXN-9AkTg6DAqAgYr9jlUABYAJm8EOuEHUAHpod4wQgfkVgCw6FQsCBSggIWgqIZyABhEJnEZxBJJUwsK43O4PdoOTqvWy5fJrAIlMpQcgAOlIUCcNMSGKkwHUSiqeAZ11u9xQzTZXiyp2GkXSeCJJPJlK13ERqKkQA)

### 🙂 Expected behavior

I don't understand why typescript is complaining about this error.

**Comment by Andarist** [comment]
It's because you can do this:
```ts
<P<{ fieldname: "b" }> callback={handleCallback} />;
```

In this scenario, your `{ fieldname: "a" }` is not assignable to this instantiation of `T`. This means that it can't be trusted within the `P`'s implementation as something that is definitely assignable to `T`.
